### PR TITLE
support record types implementing json.Marshaler

### PR DESCRIPTION
**Pull request by kr**
This works for fields, maps, and sequences.

### CODE REVIEW

File: gen.go [hunk 1]
```diff
@@ -56,34 +58,105 @@ type fieldInfo struct {
...
+			vs := spec.(*ast.ValueSpec)
+			iface := pi.Types[vs.Type].Type
+			if iface == nil {
+				return fmt.Errorf("interface assertion has no interface: %v", esprint(vs))
+			}
+			if n, ok := iface.(*types.Named); !ok || n.String() != "encoding/json.Marshaler" {
+				return fmt.Errorf("unsupported interface: %v", n)
+			}
+
+			for i, value := range vs.Values {
```
Comment (bobg) [inline gen.go:105]: Would checking for `len(vs.Names) == 1` be more appropriate than a loop here (and then refer only to `vs.Names[0]`)?

**Review by bobg — APPROVED** [review]
LGTM with one suggestion.

File: gen.go (hunk 1 above)
Comment (kr) [reply gen.go:105]: Oh referring to `vs.Names[0]` should be `vs.Names[i]`, thanks for the catch!

Comment (kr) [reply gen.go:105]: Oh sorry I misunderstood your comment. Yeah we could do it that way. This way allows

```
var _, _ json.Marshaler = (*A)(nil), (*B)(nil)
```

which, if someone really wants to write it that way for whatever reason, seems okay?

Comment (kr) [reply gen.go:105]: But I don't feel super strongly about that so I'm also down to change it if you still think it'd be better to require just one.

**Pull request merged**